lxc: Add completions for server keys and console

kadinsayani opened the pull request:
Resolves https://github.com/canonical/lxd/issues/14098.

All the improvements made to the `lxc config get/set` completions have been tested locally. I made a minor change that deviates from the requirements outlined in https://github.com/canonical/lxd/issues/14098. Instead of typing `lxc config get foo:core.<tab>` to retrieve server configuration options, typing `lxc config get foo: core.<tab>` (notice the [SPACE]) will now return server configuration options. This change was implemented to differentiate from the conventional usage of `remote:instance` across the CLI. However, please let me know if you'd prefer server configuration completions without the [SPACE] 😄.

simondeziel commented:
> I made a minor change that deviates from the requirements outlined in #14098. Instead of typing `lxc config get foo:core.<tab>` to retrieve server configuration options, typing `lxc config get foo: core.<tab>` (notice the [SPACE]) will now return server configuration options. This change was implemented to differentiate from the conventional usage of `remote:instance` across the CLI. However, please let me know if you'd prefer server configuration completions without the [SPACE] 😄.

That was the right thing to do as I got the example in the bug wrong, thanks for catching/fixing this!

kadinsayani commented on lxc/completion.go:
```diff
@@ -211,6 +212,21 @@ func (g *cmdGlobal) cmpImages(toComplete string) ([]string, cobra.ShellCompDirec
+	var keys []string
+	cmpDirectives := cobra.ShellCompDirectiveNoFileComp
+
+	if strings.Contains(instanceName, ":") {
+		instanceName = strings.Split(instanceName, ":")[1]
+
+		if instanceName == "" {
+			serverKeys, directives := g.cmpServerAllKeys()
+			keys = append(keys, serverKeys...)
+			cmpDirectives = directives
+
+			return keys, cmpDirectives
+		}
+	}
```
Thanks for the suggestion! I can see the potential for some optimization here, but we only want to modify `instanceName` if we're checking for server config keys. `instanceName` is later passed into `ParseServers()` and we want to make sure we're grabbing the right keys for an instance belonging to a remote. There may be an edge case where two distinct remotes have an instance with the same name, with one instance being a VM and one being a container (different options for config keys). 

We could use `strings.Cut()` but store the result in a variable named `remoteOnly`, or leave the code as is.

kadinsayani replied:
I'm okay with either approach 😄

markylaing commented on lxc/completion.go:
```diff
@@ -211,6 +212,21 @@ func (g *cmdGlobal) cmpImages(toComplete string) ([]string, cobra.ShellCompDirec
+	cmpDirectives := cobra.ShellCompDirectiveNoFileComp
+
+	if strings.Contains(instanceName, ":") {
+		instanceName = strings.Split(instanceName, ":")[1]
```
You can likely use `strings.Cut` here to get the instance name and check if the colon is present in one step.

markylaing commented on lxc/completion.go:
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/spf13/cobra"
 
 	"github.com/canonical/lxd/lxd/instance/instancetype"
+	"github.com/canonical/lxd/lxd/node"
```
You won't be able to import this package into the client unfortunately. That's why the build is failing.

kadinsayani commented:
I’ve updated the `cmpServerAllKeys()` function to use the metadata API by calling `GetMetadataConfiguration()`. I believe it would be more efficient to create a `MetadataConfiguration` struct in the `/shared/api` directory. This way, we can avoid parsing the response in the `completions.go` file in polynomial time. Instead, we can unmarshal the necessary information to a nested struct. I'm not sure if it would improve performance, but it would definitely be more elegant. Thoughts?

Also, @simondeziel, you were right all along about `strings.cut()` 😄.

simondeziel commented on client/lxd_server.go:
```diff
@@ -186,3 +187,50 @@ func (r *ProtocolLXD) GetMetrics() (string, error) {
+
+// GetMetadataConfiguration returns metadata configuration for a server.
+func (r *ProtocolLXD) GetMetadataConfiguration() (*map[string]any, error) {
+    // Check that the server supports it.
+	err := r.CheckExtension("metadata_configuration")
```
Tabs and spaces are mixed here.

simondeziel commented on client/interfaces.go:
```diff
@@ -73,6 +73,7 @@ type InstanceServer interface {
 	ImageServer
 
 	// Server functions
+    GetMetadataConfiguration() (config *map[string]any, err error)
```
Tabs should be used instead of spaces. Thx

tomponline replied:
Or more precisely: go fmt should be used for all code when saving in your ide

markylaing commented on client/lxd_server.go:
```diff
@@ -186,3 +187,50 @@ func (r *ProtocolLXD) GetMetrics() (string, error) {
+		return nil, err
+	}
+
+	// Prepare the request.
+	requestURL, err := r.setQueryAttributes(fmt.Sprintf("%s/1.0/metadata/configuration", r.httpBaseURL.String()))
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Send the request.
+	resp, err := r.DoHTTP(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("Bad HTTP status: %d", resp.StatusCode)
+	}
+
+	// Get the content.
+	content, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal content.
+	config := make(map[string]any)
+	err = json.Unmarshal(content, &config)
+	if err != nil {
+		return nil, err
+	}
```
I think you can use `r.queryStruct` here:
```go
config := make(map[string]any)
_, err = r.queryStruct(http.MethodGet, api.NewURL("metadata", "configuration").String(), nil, "", &config)
if err != nil {
    return err
}

return config, nil
```

Also note that this function will extract only the metadata from the request body. Your method here is returning the full API response as a `*map[string]any` which we don't need. The `queryStruct` function will return an error if the server returns an error.

Lastly, there's no need to return a `*map[string]any` as a map is already a pointer, can you please just return `map[string]any`, thanks.
```

markylaing reviewed (changes requested):
I think it's fine to create a type in the `shared/api` package for the configuration metadata. It is exposed over the API and would break the UI if changed, so we are beholden to it.

kadinsayani commented:
@markylaing I rebased my local branch on your PR branch https://github.com/canonical/lxd/pull/14132.

tomponline commented:
@kadinsayani should we also import https://github.com/lxc/incus/pull/1231 ?

kadinsayani commented:
> @markylaing I rebased my local branch on your PR branch #14132. Is this okay, or should I wait for your PR to be merged and then rebase on main?

I reverted this to avoid conflicts.

tomponline commented:
tests failing

kadinsayani commented:
> tests failing

Awaiting merge of https://github.com/canonical/lxd/pull/14132 to `latest/edge`.

kadinsayani commented:
Rebased and good to go 👍

kadinsayani commented on lxc/completion.go:
```diff
@@ -241,6 +252,38 @@ func (g *cmdGlobal) cmpInstanceAllKeys(instanceName string) ([]string, cobra.She
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	for _, field := range server {
+		for _, keyMap := range field.Keys {
+			for key := range keyMap {
+				keys = append(keys, key)
+			}
+		}
+	}
```
We can preallocate the slice:

```Go
keyCount := 0
	for _, field := range server {
		keyCount += len(field.Keys)
	}
	keys = make([]string, 0, keyCount)

	for _, field := range server {
		for _, keyMap := range field.Keys {
			for key := range keyMap {
				keys = append(keys, key)
			}
		}
	}
```

simondeziel replied:
Thanks, that seems to be a pattern used by Go libs, like in https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/strings/strings.go;l=243-250

simondeziel commented on lxc/completion.go:
```diff
@@ -241,6 +252,45 @@ func (g *cmdGlobal) cmpInstanceAllKeys(instanceName string) ([]string, cobra.She
+		keyCount += len(field.Keys)
+	}
+
+	keys = make([]string, 0, keyCount)
```
```suggestion
	keys := make([]string, 0, keyCount)
```

Maybe and remove the earlier declaration?

kadinsayani replied:
I also prefer short variable declarations in functions, thanks for pointing this out!
